core/sched.cc

```diff
@@ -289,8 +290,26 @@ void cpu::reschedule_from_interrupt(bool called_from_yield,
             return;
 #endif
         } else if (!called_from_yield) {
-            auto &t = *runqueue.begin();
-            if (p->_runtime.get_local() < t._runtime.get_local()) {
+           auto &t = *runqueue.begin();
+            if (p->_realtime._priority > 0) {
+                // Only switch to a higher-priority realtime thread, or to
+                // the next one with equal priority if yielding.
+                if (t._realtime._priority < p->_realtime._priority) {
+#ifdef __aarch64__
+                    return switch_data;
+#else
+                    return;
+#endif
+                }
+                if (t._realtime._priority == p->_realtime._priority &&
+                        !called_from_yield) {
```

Comment:
You're right, and it's even worse - I think the if of line 291/292 is not the right thing. I think I have a bug. According to https://man7.org/linux/man-pages/man2/sched_yield.2.html, "If the calling thread is the only thread in the highest priority list at that time, it will continue to run after a call to sched_yield().". So the code wrongly runs some other thread even in non-real-time priorities. I'll need to fix this.